File: 2020/13_ShuttleSearch/test_bus.py
# ======================================================================
# Shuttle Search
#   Advent of Code 2020 Day 13 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         t e s t _ b u s . p y
# ======================================================================
"Test people mover for Advent of Code 2020 day 13, Shuttle Search"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import unittest

import bus

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
EXAMPLES = [
    {'at': 929, 'bid': 7, 'departs': 931},
    {'at': 929, 'bid': 13, 'departs': 936},
    {'at': 929, 'bid': 59, 'departs': 944},
    {'at': 929, 'bid': 31, 'departs': 930},
    {'at': 929, 'bid': 19, 'departs': 931},
    {'at': 931, 'bid': 7, 'departs': 931},
    {'at': 931, 'bid': 13, 'departs': 936},
    {'at': 931, 'bid': 59, 'departs': 944},
    {'at': 931, 'bid': 19, 'departs': 931},
    {'at': 939, 'bid': 7, 'departs': 945},
    {'at': 939, 'bid': 13, 'departs': 949},
    {'at': 939, 'bid': 59, 'departs': 944},
]

# ======================================================================
#                                                              TestBus
# ======================================================================


class TestBus(unittest.TestCase):  # pylint: disable=R0904
    "Test Bus object"

    def test_empty_init(self):
        "Test the default Bus creation"

        # 1. Create default Bus object
        myobj = bus.Bus()

        # 2. Make sure it has the default values
        self.assertEqual(myobj.bid, 0)
        self.assertEqual(myobj.offset, 0)

    def test_value_init(self):
        "Test the Bus object creation from text"

        # 1. Create Bus object from text
        myobj = bus.Bus(bid='7', offset=2)

        # 2. Make sure it has the expected values
        self.assertEqual(myobj.bid, 7)
        self.assertEqual(myobj.offset, 2)

    def test_part_one(self):
        "Test part one examples of Bus object"

        # 1. Loop for all the examples
        for example in EXAMPLES:

            # 2. Create Bus object from values
            myobj = bus.Bus(bid=example['bid'])

            # 3. Check the part one result
            self.assertEqual(myobj.next_depart_after(example['at']), example['departs'])


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      t e s t _ b u s . p y                     end
# ======================================================================

File: 2020/13_ShuttleSearch/test_bus.lua
-- ======================================================================
-- Shuttle Search
--   Advent of Code 2020 Day 13 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                       t e s t _ b u s . l u a
-- ======================================================================
-- Test Bus for Advent of Code 2020 day 13, Shuttle Search

-- ----------------------------------------------------------------------
--                                                                require
-- ----------------------------------------------------------------------
local luaunit = require('luaunit')

local bus = require('bus')

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                                TestBus
-- ======================================================================

function test_empty_init()
  -- Test the default Bus creation

  -- 1. Create default Bus object
  local myobj = bus:Bus()

  -- 2. Make sure it has the default values
  luaunit.assertEquals(myobj.part2, false)
  luaunit.assertEquals(myobj.bid, 0)
  luaunit.assertEquals(myobj.offset, 0)

end

function test_text_init()
  -- Test the Bus object creation from text

  -- 1. Create Buses object from text
  local myobj = bus:Bus({bid=13, offset=1})

  -- 2. Make sure it has the expected values
  luaunit.assertEquals(myobj.bid, 13)
  luaunit.assertEquals(myobj.offset, 1)
  
  -- 3. Check methods
  luaunit.assertEquals(myobj:next_departure(929), 936)
  luaunit.assertEquals(myobj:next_departure(939), 949)

end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
os.exit( luaunit.LuaUnit.run() )

-- ======================================================================
-- end                      t e s t _ b u s . p y                     end
-- ======================================================================
